File: src/18_DeleteNodeInList/18_DeleteNodeInList.cpp
// 面试题18：给定单向链表的头指针和一个要删除的节点的值，
// 定义一个函数删除该节点。返回删除后的链表的头节点。
// 题目保证链表中节点的值互不相同

/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode(int x) : val(x), next(nullptr) {}
 * };
 */

#include <iostream>
#include <string>

#include "../Utilities/List.h"

using namespace std;

class Solution {
   public:
    // 提交时注释掉 delete 行
    ListNode* deleteNode( ListNode* head, int val ) {
        if ( !head ) return nullptr;

        ListNode *pre = head, *cur = head->next;
        // 如果要删除的节点为头节点
        if ( head->val == val ) {
            delete head;
            return cur;
        }
        // 遍历链表查找节点
        while ( cur && cur->val != val ) {
            pre = cur;
            cur = cur->next;
        }
        // 删除节点
        if ( cur ) {
            pre->next = cur->next;
            delete cur;
        }
        return head;
    }
};

class Test {
   public:
    void checkResult( const string testName, ListNode* head, int val ) {
        cout << "==== " << testName << " ====\n";
        ListNode* resultHead = solution.deleteNode( head, val );
        printList( resultHead );
        destroyList( resultHead );
        cout << endl;
    }

    // 链表中有多个结点，删除中间的结点
    void Test1() {
        ValVec vals{ 1, 2, 3, 4, 5 };
        ListNodeVec nodes = createListNodesVec( vals );
        connectListNodesVec( nodes );

        int val = 3;
        checkResult( "Test1", nodes[ 0 ], val );
    }

    // 链表中有多个结点，删除尾结点
    void Test2() {
        ValVec vals{ 1, 2, 3, 4, 5 };
        ListNodeVec nodes = createListNodesVec( vals );
        connectListNodesVec( nodes );

        int val = 5;
        checkResult( "Test2", nodes[ 0 ], val );
    }

    // 链表中有多个结点，删除头结点
    void Test3() {
        ValVec vals{ 1, 2, 3, 4, 5 };
        ListNodeVec nodes = createListNodesVec( vals );
        connectListNodesVec( nodes );

        int val = 1;
        checkResult( "Test3", nodes[ 0 ], val );
    }

    // 链表中只有一个结点，删除头结点
    void Test4() {
        ValVec vals{ 1 };
        ListNodeVec nodes = createListNodesVec( vals );
        connectListNodesVec( nodes );

        int val = 1;
        checkResult( "Test4", nodes[ 0 ], val );
    }

    // 链表为空
    void Test5() {
        int val = -1;
        checkResult( "Test5", nullptr, val );
    }

    void runTest() {
        Test1();
        Test2();
        Test3();
        Test4();
        Test5();
    }

   private:
    Solution solution;
};

int main() {
    Test test;
    test.runTest();
    return 0;
}

File: src/18_DeleteNodeInList/18_DeleteNodeInList.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, x):
#         self.val = x
#         self.next = None

import os
import sys

sys.path.append(os.path.split(sys.path[0])[0])

from Utilities import list_utils
from Utilities.list_utils import ListNode


class Solution:
    def deleteNode(self, head: ListNode, val: int) -> ListNode:
        if not head:
            return None
        if head.val == val:
            return head.next

        pre, cur = head, head.next
        while cur and cur.val != val:
            pre, cur = cur, cur.next
        if cur:
            pre.next = cur.next
        return head


class Test:
    def __init__(self) -> None:
        self.solution = Solution()

    def checkResult(self, testName: str, head: ListNode, val: int) -> None:
        print(f"==== {testName} ====")
        resultHead = self.solution.deleteNode(head, val)
        list_utils.printList(resultHead)
        print("\n")

    # 链表中有多个结点，删除中间的结点
    def Test1(self):
        vals = [1, 2, 3, 4, 5]
        nodes = list_utils.createListNodesVec(vals)
        list_utils.connectListNodesVec(nodes)

        val = 3
        self.checkResult("Test1", nodes[0], val)

    # 链表中有多个结点，删除尾结点
    def Test2(self):
        vals = [1, 2, 3, 4, 5]
        nodes = list_utils.createListNodesVec(vals)
        list_utils.connectListNodesVec(nodes)

        val = 5
        self.checkResult("Test2", nodes[0], val)

    # 链表中有多个结点，删除头结点
    def Test3(self):
        vals = [1, 2, 3, 4, 5]
        nodes = list_utils.createListNodesVec(vals)
        list_utils.connectListNodesVec(nodes)

        val = 1
        self.checkResult("Test3", nodes[0], val)

    # 链表中只有一个结点，删除头结点
    def Test4(self):
        vals = [1]
        nodes = list_utils.createListNodesVec(vals)
        list_utils.connectListNodesVec(nodes)

        val = 1
        self.checkResult("Test4", nodes[0], val)

    # 链表为空
    def Test5(self):
        val = -1
        self.checkResult("Test5", None, val)

    def runTest(self):
        self.Test1()
        self.Test2()
        self.Test3()
        self.Test4()
        self.Test5()


if __name__ == "__main__":
    test = Test()
    test.runTest()
